knimin/lib/data_access.py

```diff
@@ -2450,6 +2452,866 @@ def set_deposited_ebi(self):
+            plate_dims = dict(TRN.execute_fetchindex()[0])
+
+            # Check that the given layout has the same dimensions as the
+            # plate type
+            if plate_dims['rows'] != l_rows or plate_dims['cols'] != l_cols:
+                raise ValueError(
+                    "The given layout doesn't match the plate type "
+                    "dimensions. Plate type: (%d, %d). Layout: (%d, %d)"
+                    % (plate_dims['rows'], plate_dims['cols'], l_rows, l_cols))
+
+            # We can set the new layout. First clear the previous layout
+            # and insert the new one
+            self._clear_sample_plate_layout(sample_plate_id)
+            sql = """INSERT INTO pm.sample_plate_layout
+                        (sample_plate_id, sample_id, row, col, name, notes)
+                     VALUES (%s, %s, %s, %s, %s, %s)"""
+            sql_args = []
+            for row_idx, row in enumerate(layout):
+                for col_idx, values in enumerate(row):
+                    sql_args.append([sample_plate_id, values.get('sample_id'),
+                                     row_idx, col_idx, values.get('name'),
+                                     values.get('notes')])
+            TRN.add(sql, sql_args, many=True)
+            TRN.execute()
+
+    def read_sample_plate_layout(self, sample_plate_id):
+        """Reads the layout of a sample plate
+
+        Parameters
+        ----------
+        sample_plate_id : int
+            ID of the sample plate whose layout is to be read
+
+        Returns
+        -------
+        layout : list of list of dict
+            A 2-D matrix storing the per well information in a dict with the
+            format: {'sample_id': str, 'name': str, 'notes': str}
+        """
+        with TRN:
+            self._sample_plate_exists(sample_plate_id)
+            sql = """SELECT sample_id, col, row, name, notes
+                     FROM pm.sample_plate_layout
+                     WHERE sample_plate_id = %s
+                     ORDER BY row, col"""
+            TRN.add(sql, [sample_plate_id])
+            res = TRN.execute_fetchindex()
+            layout = []
+            row = []
+            # To construct the output layout we just need to iterate over
+            # the results since they are correctly sorted
+            for well in res:
+                # Transform the DictCursors to an actual dictionary
+                # and remove the values that we are not returning: row and col
+                well = dict(well)
+                col = well.pop('col')
+                del well['row']
+                # A new row starts when col == 0, but we need to take into
+                # account the first row, so we don't append an empty row to
+                # the layout
+                if col == 0 and row:
+                    layout.append(row)
+                    row = []
+                row.append(well)
```

Comment:
minor, but it seems like the operations involved with `well` are on lines 3140 - 3142, shouldn't this statement be grouped with those operations? no change in behavior would be impacted just on code organization